### PR TITLE
[JIT] Refactors shape logic out of code generation, fixes possible segfault

### PR DESCRIPTION
**Pull request by mruberry**
This PR:

- Removes shape logic from the code generator, which was previously relied on to return chunk and concat information
- Copies the logic to detect if a kernel has a rand_like node to the executor, making its pass independent of the code generator 
- Fixes a possible segfault where references to a vector still being modified were relied upon

The actual shape logic is unchanged.

The possible segfault is in the handling of the former "flat_inputs" in codegen.cpp. This vector holds pairs, and the second element of these pairs is a reference. In some cases these would be references to items in the vector chunk_desc, which could be added to later, possibly invalidating any references to items in it. I hit a similar segfault in testing when naively making parallel code for "flat_outputs." 

I'm submitting this small PR because it's separable, self-contained, has a fix, and I am trying to actively get away from large PRs to encourage more stability and incremental change in the fuser. 

@ngimel @zou3519

### CODE REVIEW

**Comment by mruberry** [comment]
I will likely move upfront compilation into the kernel spec constructor to deal with the asan build issue (per discussion with @ngimel). I don't love hiding functional activities like "upfront compilation" in object constructors, but it seems appropriate in this case.

**Comment by zou3519** [comment]
@mruberry asan test failure looks real

**Comment by mruberry** [comment]
Additional failures seem unrelated. Asan issue is fixed.

**Comment by zou3519** [comment]
cc @t-vi and @apaszke too. @t-vi These changes might affect how you're planning on changing the codegen

**Review by zou3519 — APPROVED** [review]
looks fine to me

File: torch/csrc/jit/fuser/codegen.cpp [hunk 1]
```diff
@@ -420,11 +368,10 @@ generateKernel(
   }
```
Comment (zou3519) [inline torch/csrc/jit/fuser/codegen.cpp]: On line 362: has_random isn't used anymore and can be deleted

Comment (mruberry) [reply torch/csrc/jit/fuser/codegen.cpp]: It's checked on line 397 to see if we should include RandHeader, RandParam, and RandInit.

Comment (zou3519) [reply torch/csrc/jit/fuser/codegen.cpp]: Oh I see, thanks for pointing that out!

**Comment by t-vi** [comment]
Thank you for the heads up!

**Pull request closed without merging**